architecture/2025-06-non-k8s-controlplane/2025-06-graph-db-replace-planes.md

```diff
@@ -0,0 +1,489 @@
+7.  **Component Updates:** Radius components that currently perform application graph operations (traversals, relationship queries) will be updated to use the GAL for graph queries, then retrieve full resource data via `database.Client` based on the graph results, while continuing to use `database.Client` directly for individual resource CRUD operations.
+
+**Storage Strategy Rationale:** We chose to maintain the existing `database.Client` interface for resource storage because: (1) The current Postgres and etcd backends already provide excellent performance for individual resource operations; (2) Graph databases are optimized for relationship queries, not necessarily single-record lookups; (3) This approach allows incremental migration with lower risk; (4) Different data types (individual resources vs. relationships) have different access patterns and requirements; (5) Production deployments can continue using battle-tested storage patterns while gaining graph capabilities for application relationship queries. **The GAL will store only minimal resource metadata (ID, type, connections, and optional query-filtering properties) in the graph database, with complete resource data retrieval continuing through the existing `database.Client` based on graph query results.**
+
+#### Architecture Diagram
+
+```mermaid
+graph TD
+    A["Radius CLI/API <br/>(User Interactions)"]
+    B["Radius CoreRP<br/>"]
+    C["database.Client <br/>(Resource Storage)"]
+    D["Graph Access Layer <br/>(Application Graph)"]
+    E1["etcd (Current) | Postgres <br/>(Production)"]
+    F1["Kùzu Embedded DB <br/>(Dev/Test/POC)"]
+    F2["Postgres + Apache AGE <br/>(Production)"]
+
+    A <--> B
+    B <-->|"resource CRUD"| C
+    B <-->|"graph queries"| D
+    C --> E1
+    D -- "dev/test/POC" --> F1
+    D -- "production" --> F2
+```
+
+* **Current (Simplified):** Radius Core Components <-> database.Client <-> etcd/Postgres
+* **Proposed:** 
+  * Resource Storage: Radius Core Components <-> database.Client <-> etcd/Postgres (unchanged)
+  * Application Graph: Radius Core Components <-> Graph Access Layer <-> Graph Database (Kùzu or Postgres+AGE)
+
+#### Detailed Design
+
+1.  **Graph Access Layer (GAL) Implementation:**
+    * The GAL will be implemented as a Go service that abstracts all application graph operations, working alongside the existing `database.Client` interface.
+    * It will provide a pluggable interface allowing different graph database backends.
+    * Configuration will determine which backend to use (Kùzu for dev/test, Postgres+AGE for production).
+    * The GAL will be responsible for maintaining synchronization between resource changes and their graph representations.
+
+2.  **Dev/Test/POC Backend - Kùzu Integration:**
+    * The Kùzu Go driver (`github.com/kuzudb/go-kuzu`) will be used.
```

Comment:
Are we planning to take a hard dependency from the control plane to Kuzu? Is there a way to make it config-driven? How about deploying Kuzu or Postgres as a container by default instead of having a direct dependency?